File: src/IKVM.Runtime/Java/Externs/sun/nio/fs/DotNetWatchService.cs
using IKVM.Runtime.Accessors.Sun.Nio.Fs;
using usnfs = IKVM.Runtime.Util.Sun.Nio.Fs;

namespace IKVM.Java.Externs.sun.nio.fs
{
    internal static partial class DotNetWatchService
    {

        public static void close0(object key) => close0Impl(key);

        public static void register0(object fs, object key, object dir, object[] events, params object[] modifiers) => register0Impl(fs, key, dir, events, modifiers);

        static partial void close0Impl(object key);

        static partial void register0Impl(object fs, object key, object dir, object[] events, object[] modifiers);

    }

#if !FIRST_PASS

    internal static partial class DotNetWatchService
    {

        private static DotNetPathAccessor dotNetPath;

        private static DotNetPathAccessor DotNetPath => JVM.BaseAccessors.Get(ref dotNetPath);

        static partial void close0Impl(object key)
            => usnfs.DotNetWatchService.Close(key);

        static partial void register0Impl(object fs, object key, object dir, object[] events, object[] modifiers)
            => usnfs.DotNetWatchService.Register(fs, key, DotNetPath.GetPath(dir), events, modifiers);

    }

#endif

}

File: src/IKVM.Java/local/sun/nio/fs/DotNetWatchService.java
package sun.nio.fs;

import java.io.Closeable;
import java.io.IOException;
import java.lang.IllegalArgumentException;
import java.nio.file.ClosedWatchServiceException;
import java.nio.file.Path;
import java.nio.file.ProviderMismatchException;
import java.nio.file.WatchEvent;
import java.nio.file.WatchKey;
import java.util.concurrent.ConcurrentHashMap;

final class DotNetWatchService extends AbstractWatchService {
    private final ConcurrentHashMap<DotNetPath, DotNetWatchKey> keys = new ConcurrentHashMap();

    @Override
    void implClose()
            throws IOException {
        for (DotNetWatchKey key : keys.values()) {
            key.close();
        }

        keys.clear();
    }

    @Override
    WatchKey register(Path path, WatchEvent.Kind<?>[] events, WatchEvent.Modifier... modifiers)
            throws IOException {
        if (!isOpen()) {
            throw new ClosedWatchServiceException();
        }
        if (!(path instanceof DotNetPath)) {
            path.getClass();
            throw new ProviderMismatchException();
        }
        final DotNetPath dir = (DotNetPath)path;

        final DotNetWatchKey key = keys.computeIfAbsent(dir, c -> new DotNetWatchKey(c));
        register0(key, dir, events, modifiers);
        return key;
    }

    void remove(DotNetPath dir) {
        final DotNetWatchKey key = keys.remove(dir);
        close0(key, dir);
    }

    final class DotNetWatchKey extends AbstractWatchKey implements Closeable {
        private final DotNetPath dir;
        private Object state;

        DotNetWatchKey(DotNetPath dir) {
            super(dir, DotNetWatchService.this);
            this.dir = dir;
        }

        @Override
        public void cancel() {
            synchronized (DotNetWatchService.this.closeLock()) {
                DotNetWatchService.this.keys.remove(dir, this);
                close();
            }
        }

        @Override
        public synchronized void close() {
            close0(this, dir);
        }

        @Override
        public boolean isValid() {
            return state != null;
        }

        void error() {
            cancel();
            signal();
        }
    }

    static native void close0(Object self, Object dir);

    static native void register0(Object self, Object dir, Object[] events, Object... modifiers);
}